### DISCUSSION TITLE
Duplo train project, how to read the color sensor

### DISCUSSION THREAD

**Discussion by thebrickmg** [pybricks/support#1372] [Q&A]
Hi
I just started to code with pybricks. I have a bit of an idea how to code, but I get stuck in this Duplo train project. 
I do build Lego GBC Modules (Great Ball Contraption) and I already have a module with the Duplo train, The problem is, that the train stops after 15 minutes. I have a video here:
https://youtu.be/VLBbY-jCAsI

So I was very happy, when I found this projekt:
https://pybricks.com/projects/tutorials/wireless/hub-to-hub/pybricks-to-official-fw/

The train should go forward and backward (not in an endless round) to go from loading to unloading. So I still have to check, were is the end of the tracks. And for this reason I need to read the color sonsor in the train. And unfortunatly I have no idea how to do that. I have no problem to use different colors and both ends. 
Thank you for any help

**Comment by laurensvalk** [comment]
@nstrijbosch, did you perhaps have an existing example to read some data from a LWP3 device?

Our example (above) only sends stuff.

**Comment by NStrijbosch** [comment]
This one worked at some point (reading the color sensor of Mario).  With some small changes this should work with the duplo train. 
 
``` python
from pybricks.hubs import ThisHub
from pybricks.iodevices import LWP3Device
from pybricks.parameters import Color

MARIO = const(0x43)
PORT_VALUE_MSG = const(0x45)

hub = ThisHub()

# Connect to the Mario
mario = LWP3Device(MARIO)
print("connected")

#subscribe to Mario barcode/color sensor
mario.write(b'\x0a\x00\x41\x01\x00\x01\x00\x00\x00\x01')

while True:
    msg = mario.read()

    kind = msg[2]
    
    if kind != PORT_VALUE_MSG:
        continue

    color = msg[6]
    
    print("color:", color)

    if color == 21:  # Mario is on red
        hub.light.on(Color.RED)
    else:
        hub.light.on(Color.BLUE)
```

**Reply by laurensvalk** [reply]
Thanks!

I have been working on cleaning up the LWP3 class lately and sharing more internal code with the Remote class.

Since we also have support for Python-based stuff in the firmware now, it's probably not too far fetched to add dedicated classes for the Duplo train and the Mario variants.

For the Duplo hub, I suppose all we need is:
- DC power out
- front light. Is there an RGB mode? Then we could make it work just like the remote light.
- color sensor input. Is there an RGB mode? Or just color index? 
- Choo Choo sound trigger.

**Reply by thebrickmg** [reply]
well, that sounds great :)
But I can not give an answer to any of the questions...

**Reply by laurensvalk** [reply]
Yeah, it was perhaps more a note to self for further investigation :smile:

The non-technical version of that was: which functionality of the train do we want/need? :smile:

**Reply by NStrijbosch** [reply]
I believe the ront light is similar to the other lights on hubs (city/technic/boost)

The motor is indeed DC power out, but you can also measure the speed

The duplo color sensor has multiple modes:
- Intensity
- Reflectivity
- Color
- RGB

The duplo speaker has two modes:
- Sound (i guess the Choo Choo, and more:   BRAKE = 3,   STATION_DEPARTURE = 5,   WATER_REFILL = 7,   HORN = 9,   STEAM = 10)
- A tone

**Reply by MimiVRC** [reply]
I'm new to this so I was curious how likely/difficult it would be to make something like this as blocks? adding sensor blocks for the mario hubs since they do support a lot of sensor features, including being able to get the scanned barcode. I know it can be done in code, but I'm looking for block versions of these features!

**Reply by dustylegobricks** [reply]
> The duplo speaker has two modes:
> 
>     * Sound (i guess the Choo Choo, and more:   BRAKE = 3,   STATION_DEPARTURE = 5,   WATER_REFILL = 7,   HORN = 9,   STEAM = 10)
> 
>     * A tone

I’ve been working directly with the Duplo Train Base over LWP3 (raw byte writes and Pybricks LWP3Device) and have confirmed that preset sounds work as expected using the documented sound IDs (3, 5, 7, 9, 10).

My question is:
Was the tone mode ever accessible over the public BLE protocol on Duplo Train Base hubs, and if so, how was it selected?

**Comment by thebrickmg** [comment]
Thank you for the fast replys. 
I guess I have to go on with trial and error with the values in the code:
`MARIO = const(0x43)

PORT_VALUE_MSG = const(0x45)
`
in the given code there is

`DUPLO_TRAIN_ID = 0x20`

 so the color sensor must be a different value, but probably not 0x43 ?
Is this correct?

**Reply by NStrijbosch** [reply]
It is the first write that subscribes to the color sensor which is important. I just did a quick test with my duplo train:

```python
from pybricks.hubs import ThisHub
from pybricks.iodevices import LWP3Device
from pybricks.parameters import Color
from pybricks.tools import wait

DUPLO_TRAIN = const(0x20)
PORT_VALUE_MSG = const(0x45)

PORT_COLOR_SENSOR = const(0x12)
MODE_RGB = const(0x03)

hub = ThisHub()

# Connect to the duplo train
train = LWP3Device(DUPLO_TRAIN)
print("connected")

#subscribe to Duplo color sensor
train.write(bytes([0x0a, 0x00, 0x41, PORT_COLOR_SENSOR, MODE_RGB, 0x01, 0x00, 0x00, 0x00, 0x01]))

wait(1000)
while True:
    msg = train.read()

    kind = msg[2]
    port = msg[3]
    
    if kind != PORT_VALUE_MSG and port != PORT_COLOR_SENSOR:
        continue

    red = msg[4]
    green = msg[6]
    blue = msg[8]
    
    print("red: ", red, "green: ", green, "blue: ", blue)
```

Hopefully this helps. 
It would be nice to add this to the Duplo_Train class which you found. But I need to spend some time to figure out how to do this properly without the read blocking everything.

**Reply by thebrickmg** [reply]
Yeah, that works with the color sensor. Now I have to find out how all will work together (train driving and color sensor reading). Thank you so far!!! Great job!!!

**Reply by thebrickmg** [reply]
This is not nice, but it works:

`
train.write(bytes([0x08, 0x00, 0x81, 0x00, 0x01, 0x51, 0x00, speed]))
`

are there any referenz values for colors like green, yellow, red and blue? 
I tried it for yellow with

- red between 180 - 220
- green between 9-30
- blue between 165-180

but the train does not stop every time driving over the yellow brick...

**Reply by Gun-neR** [reply]
> Hopefully this helps. It would be nice to add this to the Duplo_Train class which you found. But I need to spend some time to figure out how to do this properly without the read blocking everything.

Hello.  I just today picked up the Duplo Steam Train set, because it was close and cheap... Imagine my (very masculine) squeal of glee :D  Once I realised that it can be controlled by PyBricks (via a Hub).  This set went from an impulse collectable to active programmable status!

My question... Did this internal color sensor aspect get integrated into the DuploTrain() class?  I couldn't see any reference to it in this document:  https://pybricks.com/project/control-the-duplo-train/#the-full-program  Just the use of a colour sensor attached to the Huib.

**Reply by laurensvalk** [reply]
We just made a dedicated class for it. We need to update that example :smile:

**Reply by Gun-neR** [reply]
Ah, good.  Um, where can I find this class?  I have been searching all over, but aside from the aforementioned example, nada.  BTW I prefer Python over Blocks.

**Reply by BertLindeman** [reply]
I think it is imported like this:
```python
from pybricks.pupdevices import DuploTrain
```
But I am not sure from which level beta firmware.

[EDIT] And not sure for which hubs this is implemented.

**Reply by Gun-neR** [reply]
> I think it is imported like this:

Oh, that I already do.  I can control the movement, LED lighting and sound... What I was asking about was the "new" addition of reading the color sensor that is built into the train.  It is not referenced in the current duplo.py file that I have found.

**Reply by laurensvalk** [reply]
It's very new. It takes a while to document things.

**Reply by Gun-neR** [reply]
> It's very new. It takes a while to document things.

Ah, I see.  I was basing my understanding of the "time frame" according to the last early 2024 date referencing such a possible new class.  How time flies :D

**Reply by laurensvalk** [reply]
Fair point, we've been busy :smile: But while we're here, this should do it. Blocks or Python, take your pick.

You'll want the [nightly firmware](https://pybricks.com/learn/getting-started/install-pybricks/#trying-the-nightly-version-optional) to get this new class.

It's builtin, so you don't need that separate `duplo.py` anymore.

This will run on Spike or Mindstorms Inventor.

<img width="1251" height="675" alt="image" src="https://github.com/user-attachments/assets/2c071545-0f91-4395-8b8a-dd484da1186b" />

**Comment by Gun-neR** [comment]
Cool.  I had no idea how to even find or install beta firmware before.

I initially did try a City Hub, as it was what I was using prior for this train, and it seemed to have the same build-4763-gitc135fd98 ... But it couldn't import DuploTrain

So, as recommended, I switched to using a Mindstorms Hub with the following code.

I slowed it right down as my little test track is just a circle, and all five of the color plates are fairly close together.  This resulted in fairly good colour detections!  Nice!  Although the white plate was often flipping between White and Violet.  

```
from pybricks.parameters import Color
from pybricks.pupdevices import DuploTrain

# Setup
train = DuploTrain()

train.sound('depart')
train.headlights(Color.GREEN)

# Main program
while True:
    train.drive(30)
    print(train.color())
    train.headlights(train.color())
```

Anyhow, thanks for this!  I will now continue to learn me some new stuff with this impulsive purchase of a $10 Duplo Steam Train (10864) set, that I just picked up yesterday (_And I didn't even know that PyBricks was an option for it at the time, Hah!_)
